src/util/psbt/mod.rs

```diff
@@ -181,8 +185,223 @@ impl PartiallySignedTransaction {
+        self.check_index_is_within_bounds(input_index)?;
+
+        let prevouts = self.prevouts()?;
+
+        let input = &self.inputs[input_index];
+        let utxo = self.spend_utxo(input_index)?;
+        let script = utxo.script_pubkey.clone(); // scriptPubkey for input spend utxo.
+
+        if script.is_v1_p2tr() {
+            return taproot_sighash(input, prevouts, input_index, cache, tapleaf_hash);
+        }
+
+        let hash_ty = input
+            .sighash_type
+            .map(|ty| ty.ecdsa_hash_ty())
+            .unwrap_or(Ok(EcdsaSighashType::All))
+            .map_err(|_| SighashError::InvalidSighashType)?; // Only support standard sighash types.
+
+        let is_wpkh = script.is_v0_p2wpkh();
+        let is_wsh = script.is_v0_p2wsh();
+
+        let is_nested_wpkh = script.is_p2sh()
+            && input.redeem_script.as_ref()
+            .map(|s| s.is_v0_p2wpkh()).unwrap_or(false);
+
+        let is_nested_wsh = script.is_p2sh()
+            && input.redeem_script.as_ref()
+            .map(|x| x.is_v0_p2wsh()).unwrap_or(false);
+
+        let is_segwit = is_wpkh || is_wsh || is_nested_wpkh || is_nested_wsh;
+
+        let sighash = if is_segwit {
+            if is_wpkh || is_nested_wpkh {
+                let script_code = if is_wpkh {
+                    Script::p2wpkh_script_code(&script).ok_or(SighashError::NotWpkh)?
+                } else {
+                    Script::p2wpkh_script_code(input.redeem_script.as_ref().expect("checked above"))
+                        .ok_or(SighashError::NotWpkh)?
+                };
+                cache.segwit_signature_hash(input_index, &script_code, utxo.value, hash_ty)?
+
+            } else {
+                let script_code = input.witness_script.as_ref().ok_or(SighashError::MissingWitnessScript)?;
+                cache.segwit_signature_hash(input_index, script_code, utxo.value, hash_ty)?
+            }
+        } else {
+            let script_code = if script.is_p2sh() {
+                input.redeem_script.as_ref().ok_or(SighashError::MissingRedeemScript)?
+            } else {
+                &script
+            };
+            cache.legacy_signature_hash(input_index, script_code, hash_ty.to_u32())?
+        };
+
+        Ok((Message::from_slice(&sighash).expect("sighashes are 32 bytes"), hash_ty.into()))
+    }
+
+    /// Returns the prevouts for this PSBT.
+    fn prevouts<'a>(&self) -> Result<Vec<&TxOut>, SighashError> {
+        let len = self.inputs.len();
+        let mut utxos = Vec::with_capacity(len);
+
+        for i in 0..len {
+            utxos.push(self.spend_utxo(i)?);
+        }
+
+        Ok(utxos)
+    }
+
+    /// Returns the spending utxo for this PSBT's input at `input_index`.
+    fn spend_utxo(&self, input_index: usize) -> Result<&TxOut, SighashError> {
+        let input = &self.inputs[input_index];
+        // FIXME: Should we be checking the non-witness utxo first as the comment says in bdk?
```

Comment:
bip 174 spec lists that check first under "[Data Signers Check For](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#data-signers-check-for)" but I don't see a reason it's strictly necessary as the "[Always try first with the non-witness utxo](https://github.com/bitcoindevkit/bdk/blob/6e8744d59d07333205cd29c1df7b26892a3cfb60/src/wallet/signer.rs#L561)" bdk comment suggests. I suppose if more non-witness utxos get spent,  it's just a slight optimization